scripts/build_setup/build_setup.js

```diff
@@ -0,0 +1,63 @@
+const LinkerConfig = require('./linkerConfig');
+const AssetAssembly = require('./assets_asm');
+
+const fs = require('fs');
+
+const assetsDir = './assets';
+const linkerConfigFileName = './build/gametank-2M.cfg';
+const bankMakeListFileName = './build/bankMakeList.inc';
+const bankNumHeaderFileName = './src/gen/bank_nums.h';
+const modulesHeaderFileName = './src/gen/modules_enabled.h';
+const modulesAsmFileName = './src/gen/modules_enabled.inc';
+
+const projectConfig = JSON.parse(fs.readFileSync('./project.json'));
+const requiredProjectKeys = [
+    'title',
+    'romname',
+    'progbanks',
+    'modules',
+];
+
+let hadError = false;
+requiredProjectKeys.forEach((reqKey) => {
+    if(!Object.hasOwn(projectConfig, reqKey)) {
+        console.error(`Project config missing ${reqKey}`);
+        hadError = true;
+    }
+});
+if(hadError) {
+    throw new Error('Project config was missing one or more required properties');
+}
+
+const dirInfo = fs.readdirSync(assetsDir).filter((p) => fs.statSync(assetsDir + '/' + p).isDirectory());
+
+const buildCfg = LinkerConfig.generateLinkerConfig(dirInfo, projectConfig.progbanks);
+
+fs.writeFileSync(linkerConfigFileName, buildCfg.linker);
+fs.writeFileSync(bankMakeListFileName, buildCfg.bankMakeList);
+
+AssetAssembly.generateAssetAssemblyFiles(dirInfo, buildCfg.folderBankMap);
+
+fs.writeFileSync(bankNumHeaderFileName, 
+    `//@generated
+//Editing this manually is not recommended
+//This file contains bank numbers for segments not associated with asset folders
+${buildCfg.bankNumbers.map((bank) => `#define BANK_${bank.name} ${bank.num}`).join('\n')}
+`
+);
+
+fs.writeFileSync(modulesHeaderFileName, 
+    `//@generated
+//Editing this manually is not recommended
+//This file tells the SDK optional features which ones are enabled
+${projectConfig.modules.map((mod) => `#define ENABLE_MODULE_${mod}`).join('\n')}
+`
+);

```

Comment:
I don't really like how optional features are enabled/ disabled by custom syntax in this json file. I do appreciate that this approach does succeed in keeping the C and asm synced, as if that wasn't a concern just having users manually edit their `modules_enabled.h` would be much nicer (imo).

I struggled for a bit to think of an approach that wouldn't require either keeping a c defs file and and asm defs file in sync or a bespoke format to produce generated files.  On thing that came to mind was defining something in the makefile like `ENABLED_MODULES = ...` and then using something like `sed` to prepare it to be appended to `CFLAGS` and `AFLAGS` e.g.:

```
ENABLED_MODULES ?= "PERSIST RANDOM TEXT"

CFLAGS += $(shell echo $(ENABLED_MODULES) | sed s/[^" "]*/-DENABLE_MODULE_\&/g)
AFLAGS += $(shell echo $(ENABLED_MODULES) | sed s/[^" "]*/-D" "ENABLE_MODULE_\&/g)
```

Although this solution feels similarly janky and ad hoc.  I'm curious to what your thoughts are

Reply:
Nit: if you do decide to keep the above, I think `//This file describes which optional features of the SDK are enabled` would be a bit clearer

Reply:
This above solution would also mean that it isn't necessary to include `modules_enabled.h` or `modules_enabled.inc` in each of the optional modules

Reply:
One thing about the solution that uses the `modules_enabled.h` file is that VSCode knows easily whether to gray out the syntax between the ifdefs, without having to get it to understand flags passed to cc65 in the makefile.

NodeJS projects keep these kind of configs in a `package.json` fairly often, and I think one particular thing it does well is separating configuration from the build mechanics. At work I'm constantly told "oh you just change that in so-and-so file" and then have to go swim through a thirty page makefile and eventually decode which was meant to be the configurable part. The `project.json` here is meant to be a file that only contains things it is okay to edit.